File: Colivi-frontend/src/features/housing/api/accommodationService.ts
import api from '../../../lib/api';
import type {
  AccommodationRequest,
  AccommodationResponse,
  AccommodationVisibility,
  Page,
} from '../types/accommodation.types';

// ── Accommodation Service ─────────────────────────────────────────

export const accommodationService = {
  /**
   * POST /api/v1/accommodation
   * Creates a new physical accommodation owned by the authenticated user.
   */
  create: async (data: AccommodationRequest): Promise<AccommodationResponse> => {
    const response = await api.post<AccommodationResponse>('/accommodation', data);
    return response.data;
  },

  /**
   * GET /api/v1/accommodation/me
   * Returns paginated list of accommodations owned by the authenticated user.
   */
  getMyAccommodations: async (
    visibility: AccommodationVisibility = 'AVAILABLE',
    page = 0,
    size = 10,
  ): Promise<Page<AccommodationResponse>> => {
    const response = await api.get<Page<AccommodationResponse>>('/accommodation/me', {
      params: { visibility, page, size },
    });
    return response.data;
  },

  /**
   * GET /api/v1/accommodation/:id
   * Returns a single accommodation by ID.
   */
  getById: async (id: string): Promise<AccommodationResponse> => {
    const response = await api.get<AccommodationResponse>(`/accommodation/${id}`);
    return response.data;
  },

  /**
   * PUT /api/v1/accommodation/:id
   * Updates an existing accommodation.
   */
  update: async (id: string, data: AccommodationRequest): Promise<AccommodationResponse> => {
    const response = await api.put<AccommodationResponse>(`/accommodation/${id}`, data);
    return response.data;
  },

  /**
   * POST /api/v1/accommodation/:id/images
   * Uploads a single image for an accommodation (multipart/form-data).
   */
  uploadImage: async (id: string, file: File): Promise<AccommodationResponse> => {
    const formData = new FormData();
    formData.append('file', file);
    const response = await api.post<AccommodationResponse>(
      `/accommodation/${id}/images`,
      formData,
      { headers: { 'Content-Type': 'multipart/form-data' } },
    );
    return response.data;
  },

  /**
   * DELETE /api/v1/accommodation/:id/images/:imageId
   * Removes a specific image from an accommodation.
   */
  deleteImage: async (id: string, imageId: string): Promise<void> => {
    await api.delete(`/accommodation/${id}/images/${imageId}`);
  },

  /**
   * PATCH /api/v1/accommodation/delete/:id
   * Soft-deletes an accommodation (owner only).
   */
  softDelete: async (id: string): Promise<AccommodationResponse> => {
    const response = await api.patch<AccommodationResponse>(`/accommodation/delete/${id}`);
    return response.data;
  },
};

File: scripts/apiTester/services/accommodation_service.py
from typing import Dict, Any, List
import requests
from core.client import ApiClient

class AccommodationService:
    def __init__(self, client: ApiClient):
        self.client = client
        self.base_path = "/api/v1/accommodation"
        
    def create(self, payload: Dict[str, Any], token: str) -> Dict[str, Any]:
        response = self.client.request("POST", self.base_path, "Creación de Alojamiento", expected_status=201, token=token, json=payload)
        return response.json()
        
    def get_details(self, accommodation_id: str, token: str) -> Dict[str, Any]:
        response = self.client.request("GET", f"{self.base_path}/{accommodation_id}", "Obtener Detalles del Alojamiento", expected_status=200, token=token)
        return response.json()
        
    def get_catalog(self, owner_id: str, token: str) -> Dict[str, Any]:
        params = {
            "ownerId": owner_id,
            "visibility": "AVAILABLE",
            "page": 0,
            "size": 10
        }
        response = self.client.request("GET", self.base_path, "Obtener Catálogo de Alojamientos", expected_status=200, token=token, params=params)
        return response.json()
        
    def update(self, accommodation_id: str, payload: Dict[str, Any], token: str) -> Dict[str, Any]:
        response = self.client.request("PUT", f"{self.base_path}/{accommodation_id}", "Actualizar Alojamiento", expected_status=200, token=token, json=payload)
        return response.json()
        
    def upload_image(self, accommodation_id: str, file_name: str, file_bytes: bytes, step_name: str, token: str) -> Dict[str, Any]:
        url = self.client._build_url(f"{self.base_path}/{accommodation_id}/images")
        files = {
            "file": (file_name, file_bytes, "image/png")
        }
        headers = {"Authorization": f"Bearer {token}"}
        
        print(f"\n--- PASO: {step_name} ---")
        try:
            response = requests.post(url, files=files, headers=headers)
        except requests.RequestException as e:
            print(f"[-] ERROR: Petición fallida ({step_name}): {e}")
            import sys
            sys.exit(1)
            
        self.client.check_response(response, 200)
        return response.json()
        
    def reorder_images(self, accommodation_id: str, payload: List[Dict[str, Any]], token: str) -> Dict[str, Any]:
        response = self.client.request("PUT", f"{self.base_path}/{accommodation_id}/images/order", "Reordenar Imágenes", expected_status=200, token=token, json=payload)
        return response.json()
        
    def delete_image(self, accommodation_id: str, image_id: str, token: str) -> None:
        self.client.request("DELETE", f"{self.base_path}/{accommodation_id}/images/{image_id}", "Eliminar Imagen del Alojamiento", expected_status=204, token=token)
        
    def soft_delete(self, accommodation_id: str, token: str) -> Dict[str, Any]:
        response = self.client.request("PATCH", f"{self.base_path}/delete/{accommodation_id}", "Soft Delete Alojamiento", expected_status=200, token=token)
        return response.json()
        
    def hard_delete(self, accommodation_id: str, admin_token: str) -> None:
        self.client.request("DELETE", f"{self.base_path}/hardDelete/{accommodation_id}", "Admin: Hard Delete Alojamiento (Accommodation)", expected_status=204, token=admin_token)
